### ISSUE TITLE
Is this ZClip support AMP training?

### ISSUE THREAD

**Issue by JohnHerry** [bluorion-com/ZClip#6]
Thanks for the job.  I am traning a deep flow-matching with AMP mode, the float16 model will make the training unsable that easily lead to grad nan or infinite grade value.  Now the ZClip may be help. so I want to know if this ZClip support the Mixed precision training？ Can I directly replace the torch.GradScaler with ZClip directly for use？   
thank you

**Comment by akanyaani** [comment]
Hi @JohnHerry Great question! Yes, ZClip is fully compatible with mixed precision (AMP) training.

ZClip is implemented to run inside the **on_before_optimizer_step** callback, which makes it well-suited for AMP workflows—especially if you're using PyTorch Lightning. In Lightning, torch.cuda.amp.GradScaler.unscale_() is automatically called before this callback, so ZClip operates on unscaled gradients, just as intended. This ensures safe and effective adaptive clipping without interfering with AMP’s scaling logic.

So to clarify:

You should continue using GradScaler for AMP as usual.

ZClip inside **on_before_optimizer_step** will work seamlessly, since Lightning handles unscaling before the callback.

After clipping, the training proceeds with scaler.step(optimizer) and scaler.update().

⚠️ If you’re not using PyTorch Lightning, just make sure that scaler.unscale_(optimizer) is called before ZClip is applied. If you're unsure how this is handled in your training loop, I’d be happy to help review it.

Let me know your setup if you're using a custom loop

**Comment by JohnHerry** [comment]
> Hi [@JohnHerry](https://github.com/JohnHerry) Great question! Yes, ZClip is fully compatible with mixed precision (AMP) training.
> 
> ZClip is implemented to run inside the **on_before_optimizer_step** callback, which makes it well-suited for AMP workflows—especially if you're using PyTorch Lightning. In Lightning, torch.cuda.amp.GradScaler.unscale_() is automatically called before this callback, so ZClip operates on unscaled gradients, just as intended. This ensures safe and effective adaptive clipping without interfering with AMP’s scaling logic.
> 
> So to clarify:
> 
> You should continue using GradScaler for AMP as usual.
> 
> ZClip inside **on_before_optimizer_step** will work seamlessly, since Lightning handles unscaling before the callback.
> 
> After clipping, the training proceeds with scaler.step(optimizer) and scaler.update().
> 
> ⚠️ If you’re not using PyTorch Lightning, just make sure that scaler.unscale_(optimizer) is called before ZClip is applied. If you're unsure how this is handled in your training loop, I’d be happy to help review it.
> 
> Let me know your setup if you're using a custom loop

Thank you very much for the kindly help. I will have a try.

**Comment by JohnHerry** [comment]
Here is a suggestion to support pure pytorch training, without lightning.    
  the `zclip.step()` can be viewed as an advanced subtitute of  `torch.nn.utils.clip_grad_norm_` which support only fixed clip value.  but the `clip_grad_norm_()` will return a grad_norm of type tensor, while the `ZClip.step()` return a grad_norm of type scalar,  that makes the replacement in existing codes `clip_grad_norm_` not so elegant.

**Comment by JohnHerry** [comment]
My instance of replace the fixed grad_norm with ZClip dynamic, did not get stable training performance.   
The original code is :  
```
model.step()
grad_norm = model.get_global_grad_norm()
if batch_idx % accumulate_grad_steps == 0:
    scaler.unscale_(optimizer)
   if torch.isfinite(grad_norm):
       scaler.step(optimizer)
   scaler.update()
   optimizer.zero_grad()
   scheduler.step()
```    
The code with ZClip is:   
```
model.step()
grad_norm = model.get_global_grad_norm()
if batch_idx % accumulate_grad_steps == 0:
    scaler.unscale_(optimizer)
    grad_norm = zclip_obj.step(model)
   if torch.isfinite(torch.tensor(grad_norm)):
       scaler.step(optimizer)
   scaler.update()
   optimizer.zero_grad()
   scheduler.step()
```
The amp training process got  Nan grad after about 90k steps.  Now I am checking the training data, maybe there are too much noisy sampels there.

**Comment by akanyaani** [comment]
@JohnHerry Could you please share the loss and gradient-norm plots from the run where you first saw NaNs? Even a simple log of gradient norms over time will help us pinpoint exactly when things go off the rails. Also, it would be really helpful to know what kind of model you’re training (architecture, size, special layers, etc.) and what your data looks like (dataset source, preprocessing steps, any known noisy samples). With those details, we’ll be much better equipped to diagnose the issue. Thanks!

**Comment by JohnHerry** [comment]
sure,  the  loss and training plot are listed below.

![Image](https://github.com/user-attachments/assets/bb078a39-bc1b-45a6-a765-b15cf6c20bb5)
![Image](https://github.com/user-attachments/assets/0888312f-9cb8-4ae5-8b44-e893119c69dc)
the model is a flow-matching work as speech token decoder [to melspectorgram],  code is opened there: 
[CosyVoice/flow](https://github.com/FunAudioLLM/CosyVoice)

**Comment by JohnHerry** [comment]
It seems that ZClip training pipeline runs faster then the fixed grad_clip value,  it cost less time each training step. but in the view of covergent speed, the ZClip makes slower training, At the same training step, the ZClip version got higher training loss then the fix-grad_clip version.

**Comment by JohnHerry** [comment]
Feed back:    
The ZClip do makes my AMP training of the flow model more stable, it can run more steps. but as to the result performance, the ZClip trained result performs a little lower then the original  fixed value cliped model - not too much.   
It is recommended if the AMP training is unstable.   
 thank you.
